lib/curl/rest.nit

```diff
@@ -0,0 +1,114 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+module rest
+
+import curl
+import native_curl
+import json
+import json::dynamic
+import json::serialization_write
+
+# Restful JSON request builder over HTTP
+#
+class CurlRestRequest
+	super CurlHTTPRequest
+
+	# Set a serializable object to be sent as a JSON string to the server
+	#
+	# Object must subclass Serializable and implement its services
+	# or being annotated with serialize. For further details :
+	# https://github.com/nitlang/nit/tree/master/lib/json
+	var json_src : nullable Serializable is writable
+
+	# Execute HTTP request
+	#
+	# Returns a CurlRestResponseSuccess as CurlResponse
+	redef fun execute : CurlResponse
+	do
+		# BASIC CURL STUFF (Reveiver, callback, user agent, URL, Address family)
+		self.curl.native = new NativeCurl.easy_init
+		var success_response = new CurlRestResponseSuccess
+
+		if not self.curl.is_ok then return answer_failure(0, "Curl instance is not correctly initialized")
+
+		var callback_receiver : CurlCallbacks = success_response
+		if self.delegate != null then callback_receiver = self.delegate.as(not null)
+
+		var err : CURLCode
+
+		# Prepare request
+		err = prepare_request(callback_receiver)
+		if not err.is_ok then return answer_failure(err.to_i, err.to_s)
+
+		# ON THE ROAD AGAIN
+		var err_resp = perform
+		if err_resp != null then return err_resp
+
+		var st_code = self.curl.native.easy_getinfo_long(new CURLInfoLong.response_code)
+		if not st_code == null then success_response.status_code = st_code
+
+		self.curl.native.easy_clean
+
+		return success_response
+	end
+
+	# Redefines body creation function. A call to this method requires a
+	# valid object in the attribute json_src otherwise it will raise an
+	# error !
+	redef protected fun set_body : CURLCode
+	do
+		if json_src!=null then
+			var wr = new StringWriter
+			var serializer = new JsonSerializer(wr)
+			serializer.plain_json=true
+			serializer.pretty_json = true
+			serializer.serialize(json_src)
+			var body = wr.to_s
+			assert body!=""
+			return self.curl.native.easy_setopt(new CURLOption.postfields, body)
+		else
+			return new CURLCode.ok
+		end
+	end
+
+end
+
+# Success response class of a restful request
+class CurlRestResponseSuccess
+	super CurlResponseSuccessIntern
+
+	# Response code from the server
+	var status_code = 0
+
+	# Raw JSON string, as received from the server
+	var json_string = ""
+
+	redef fun body_callback(line) do self.json_string += line
+
+	var json_object: nullable Object = null
+
+	# `JsonValue` to access the data in `json_string`
+	fun json_value: JsonValue do return new JsonValue(json_object)
+
+	# Deserialized Nit object from `json_string`
+	fun deserialize(type_name :String) : nullable Object
+	do
+		var deserializer = new JsonDeserializer(json_string)
+		return deserializer.deserialize(type_name)
+	end
```

Comment:
You're losing the errors here (from `deserializer.errors`). In fact, this method is not that useful as the clients could simply call `json_string.from_json_string` and get pretty much the same result.

To make this service worth it, as I suggested yesterday, move the deserialization at the end of `execute` so you can catch the errors and avoid returning a "success" response when receiving invalid JSON.

Reply:
Yeah, maybe it would be better. You bring up two good points, especially now that you moved up the `method` service, I wonder how useful is the REST request class.

1. REST is not restricted to JSON in anyway. However, your REST request only adds logic to deal with JSON. Has the class lost its focus? At least, the class should be renamed to be specificly about JSON and not REST.

2. Now that you can use your improved `CurlHTTPRequest`, is the REST/JSON request still really useful? Client can serialize to and from JSON using the 3 oneliners: `Serializable::serialize_to_json`, `String::from_json_string` or `String::to_json_value`, and have more control in order to adapt their code to a remote server/API.

You should still keep (but tweak) your example to document how to use the `curl` and `json` packages to communicate with a RESTful server in JSON. It will be useful to future users!

Reply:
Indeed REST isn't restricted to JSON and someone would have to extend this module. Maybe by turning it into a generic class ? As for now, I think this class remains useful if you consider the lazy dev point of view (I love when a lib does my job). Giving users ways to write quick script remains the aim of this PR ;)